wrappers/pkcs11/pkcs11_client.go

```diff
@@ -0,0 +1,553 @@
+// Copyright (c) 2024 OpenBao a Series of LF Projects, LLC
+// SPDX-License-Identifier: MPL-2.0
+
+package pkcs11
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"encoding/binary"
+	"encoding/hex"
+
+	"github.com/openbao/openbao/api/v2"
+	uuid "github.com/hashicorp/go-uuid"
+	pkcs11 "github.com/miekg/pkcs11"
+	wrapping "github.com/openbao/go-kms-wrapping/v2"
+)
+
+type Pkcs11Key struct {
+	label		string
+	id			string
+}
+func (k Pkcs11Key) String() string {
+	return fmt.Sprintf("%s:%s", k.label, k.id)
+}
+func newPkcs11Key(v string) (*Pkcs11Key, error) {
+	pos := strings.LastIndex(v, ":")
+	if pos <= 0 {
+		return nil, fmt.Errorf("Invalid key format")
+	}
+	k := &Pkcs11Key{
+		label: v[:pos],
+		id: v[pos+1:],
+	}
+	return k, nil
+}
+func (k Pkcs11Key) Set(v string) error {
+	pos := strings.LastIndex(v, ":")
+	if pos <= 0 {
+		return fmt.Errorf("Invalid key format")
+	}
+	k.label = v[:pos]
+	k.id = v[pos+1:]
+	return nil
+}
+
+type pkcs11ClientEncryptor interface {
+	Close()
+	Encrypt(plaintext []byte) (ciphertext []byte, keyId *Pkcs11Key, err error)
+	Decrypt(ciphertext []byte, keyId *Pkcs11Key) (plaintext []byte, err error)
+}
+
+type Pkcs11Client struct {
+	client	  	*pkcs11.Ctx
+	lib	    	string
+	slot      	uint
+	tokenLabel	string
+	pin       	string
+	keyLabel	string
+	keyId     	string
+	mechanism 	uint
+}
+
+const (
+	EnvHsmWrapperLib   		= "BAO_HSM_LIB"
+	EnvHsmWrapperSlot  		= "BAO_HSM_SLOT"
+	EnvHsmWrapperTokenLabel = "BAO_HSM_TOKEN_LABEL"
+	EnvHsmWrapperPin		= "BAO_HSM_PIN"
+	EnvHsmWrapperKeyLabel	= "BAO_HSM_KEY_LABEL"
+	EnvHsmWrapperKeyId		= "BAO_HSM_KEY_ID"
+	EnvHsmWrapperMechanism  = "BAO_HSM_MECHANISM"
+)
+
+func newPkcs11Client(opts *options) (*Pkcs11Client, *wrapping.WrapperConfig, error) {
+	var lib, slot, keyId, tokenLabel, pin, keyLabel, mechanism string
+	var slotNum, mechanismNum uint64
+	var err error
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperLib) != "" && !opts.Options.WithDisallowEnvVars:
+		lib = api.ReadBaoVariable(EnvHsmWrapperLib)
+	case opts.withLib != "":
+		lib = opts.withLib
+	default:
+		return nil, nil, fmt.Errorf("lib is required")
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperSlot) != "" && !opts.Options.WithDisallowEnvVars:
+		slot = api.ReadBaoVariable(EnvHsmWrapperSlot)
+	case opts.withSlot != "":
+		slot = opts.withSlot
+	default:
+		slot = ""
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperTokenLabel) != "" && !opts.Options.WithDisallowEnvVars:
+		tokenLabel = api.ReadBaoVariable(EnvHsmWrapperTokenLabel)
+	case opts.withTokenLabel != "":
+		tokenLabel = opts.withTokenLabel
+	default:
+		tokenLabel = ""
+	}
+
+	if slot == "" && tokenLabel == "" {
+		return nil, nil, fmt.Errorf("slot or token label required")
+	}
+
+	switch {
+	case api.ReadBaoVariable(EnvHsmWrapperKeyId) != "" && !opts.Options.WithDisallowEnvVars:
+		keyId = api.ReadBaoVariable(EnvHsmWrapperKeyId)
+	case opts.withKeyId != "":
+		keyId = opts.withKeyId
```

Comment:
I think this comment might still apply? https://github.com/openbao/go-kms-wrapping/pull/8/files/45ad5ab66ef39ae498a8a736b1a190d34ca1c397#r1718338953